### ISSUE TITLE
Before Feature Teardown

### ISSUE THREAD

**Issue by davidsteinsland** [Behat/Behat#950]
As the name suggest, I thought perhaps this event would contain the context objects, before they were destroyed. This does not seem to be the case.

Is there a way one can get the context objects, just after a feature?

I know I can work around this issue by listening on `afterScenario`, but then I need to listen for `afterScenarioOutline` also in order to catch all cases. I also need to perform extra checks to find out whether it's the last scenario/outline in the given feature.

Any thoughts on this?

**Comment by everzet** [comment]
There is an `AfterFeature` hook. However, having contexts in that hook doesn't make much sense because of how context isolation works - each scenario gets its own set of context instances and those aren't shared across scenarios.

**Comment by davidsteinsland** [comment]
The hook gets called statically and is within a Feature Context. I'm talking about the event dispatcher system.

If you read between the lines, I'm actually proposing a new event around the lines of "Last scenario in feature".

Because this just seem a bit overkill:

```php
    /* OutlineTested::AFTER + ScenarioTested::AFTER */
    public function afterFeature(AfterTested $event)
    {
        $scenario = null;

        if ($event instanceof Event\AfterScenarioTested) {
            $scenario = $event->getScenario();
        } elseif ($event instanceof Event\AfterOutlineTested) {
            $scenario = $event->getOutline();
        } else {
            return;
        }

        $environment = $event->getEnvironment();
        $feature = $event->getFeature();

        $scenarios = $feature->getScenarios();
        $lastScenario = $scenarios[count($scenarios) - 1];

        if ($lastScenario !== $scenario) {
            return;
        }
        /* last scenario in feature here */
    }
```
... especially if done in multiple event subscribers.

**Comment by everzet** [comment]
What's your actual use case?

**Comment by davidsteinsland** [comment]
My use case is actually both "Last scenario/outline in feature" and "First scenario/outline in feature", because I need to perform some extra setups and teardowns.

It is perhaps not very representative for the rest of the community, but before features we load DB seeds based on annotations in the feature. To ensure we don't mess up the application in this phase we close any browser windows (in case tabs are opened and loading ajax or something).

So basically:
Before feature: Load DB seeds/reset DB
After last scenario in feature: Close browser (needs Context instances here)
After feature: Save resulting DB seed

I could perhaps come up with other use cases as well, but this is what we're working on now.

**Comment by everzet** [comment]
I'd suggest you to rethink how you isolate your environments. Your optimisation would break as soon as you use `--order` option, as it should. I'm afraid for this reason we can't support this particular feature in the core. I'm not sure you should use this feature either, but that's your choice.
